networkit/cpp/io/EdgeListReader.cpp

```diff
@@ -10,229 +10,142 @@
+
+    auto skipWhitespaceAndSeparator = [this, &it, &end]() -> void {
+        while (it != end && (*it == ' ' || *it == separator))
+            ++it;
+    };
+
+    auto scanId = [&it]() -> node {
+        char *past;
+        auto value = strtol(it, &past, 10);
+        if (past <= it)
+            throw std::runtime_error("Error in parsing file - looking for nodeId failed");
+        it = past;
+        return value;
+    };
+
+    auto scanWeight = [&it]() -> edgeweight {
+        char *past;
+        auto value = strtod(it, &past);
+        if (past <= it)
+            throw std::runtime_error("Error in parsing file - looking for weight failed");
+        it = past;
+        return value;
+    };
+
+    // Returns
+    // 0 iff it does not point to a line ending
+    // 1 iff it is a single-char ending ('\r' or '\n')
+    // 2 iff it is two-char ending ("\r\n")
+    auto detectLineEnding = [&it, &end]() -> size_t {
+        if (*it == '\r') {
+            if ((it + 1) != end && *(it + 1) == '\n')
+                return 2;
+            return 1;
         }
-    }
-    file.close();
-    maxNode = maxNode - this->firstNode + 1;
-    DEBUG("max. node id found: " , maxNode);
-
-    Graph G(maxNode, weighted, directed);
-
-    DEBUG("second pass");
-    file.open(path);
-    // split the line into start and end node. since the edges are sorted, the start node has the highest id of all nodes
-    i = 0; // count lines
-    while(std::getline(file,line)){
-        if (line.empty()) continue;
-        if(*line.rbegin() == '\r') line.pop_back();
-        ++i;
-        if (line.compare(0, this->commentPrefix.length(), this->commentPrefix) == 0) {
-        } else {
-            std::vector<std::string> split = Aux::StringTools::split(line, this->separator);
-            if (split.size() == 2) {
-                node u = std::stoul(split[0]) - this->firstNode;
-                node v = std::stoul(split[1]) - this->firstNode;
-                if (!G.hasEdge(u,v)) {
-                    G.addEdge(u, v);
-                }
-            } else if (weighted && split.size() == 3) {
-                node u = std::stoul(split[0]) - this->firstNode;
-                node v = std::stoul(split[1]) - this->firstNode;
-                double weight = std::stod(split[2]);
-                if (!G.hasEdge(u,v)) {
-                    G.addEdge(u, v, weight);
-                }
-            } else {
-                std::stringstream message;
-                message << "malformed line ";
-                message << i << ": ";
-                message << line;
-                throw std::runtime_error(message.str());
-            }
+        if (*it == '\n')
+            return 1;
+
+        return 0;
+    };
+
+    // Map nodes and increase graph size
+    auto mapNode = [&](node in) -> node {
+        if (!continuous) {
+            // mapNodeIds is <string, node>
+            auto it = mapNodeIds.find(std::to_string(in));
```

Comment:
This doesn't make sense. Why do we read the node IDs as numbers and then convert to strings? Was the old code able to handle non-numerical node IDs?

Reply:
AFAICT, it wasn't. It stored the node mapping as `<string, node>` (https://github.com/networkit/networkit/blob/master/networkit/cpp/io/EdgeListReader.cpp#L178) but converted the IDs to `node` when adding the edges, e.g: https://github.com/networkit/networkit/blob/master/networkit/cpp/io/EdgeListReader.cpp#L210. 

I went along with this so as not to change the interface of the `getNodeMap()` function.

Reply:
Yes, the old code handled non-numerical node ids. For example the file
```
a b
a c
c d
foo bar
```

gives the node map `{'a': 0, 'b': 1, 'bar': 5, 'c': 2, 'd': 3, 'foo': 4}` and a graph with 6 nodes and 4 edges as expected.

Reply:
Let's continue to support this use case and add a test for this. (Note that this leaves some strange edge cases, e.g., nodes 01 is not equal to node 1 but hopefully does that anyway...)

Reply:
The refactored reader now supports non-numerical node IDs like before. I have added a unit test (and an edge list) for said use case.